File: packages/adapter/src/otlp/mappers/traces.rs
use crate::otlp::error::OtlpError;
use crate::otlp::values::{key_values_to_json, service_name_from_resource};
use domain::{
    normalize_span_id_bytes, normalize_trace_id_bytes, SpanRecord,
};
use opentelemetry_proto::tonic::collector::trace::v1::ExportTraceServiceRequest;
use opentelemetry_proto::tonic::trace::v1::{Span, Status};
use serde_json::json;

pub fn map_traces(request: ExportTraceServiceRequest) -> Result<Vec<SpanRecord>, OtlpError> {
    let mut out = Vec::new();

    for resource_spans in request.resource_spans {
        let resource = resource_spans.resource.as_ref();
        let resource_attributes = resource
            .map(|r| key_values_to_json(&r.attributes))
            .unwrap_or_else(|| json!([]));
        let resource_dropped = resource.map(|r| r.dropped_attributes_count).unwrap_or(0);
        let resource_schema_url = None;
        let service_name = resource
            .map(|r| service_name_from_resource(&r.attributes))
            .unwrap_or_else(|| "unknown_service".into());

        for scope_spans in resource_spans.scope_spans {
            let scope = scope_spans.scope.as_ref();
            let scope_name = scope.map(|s| s.name.clone()).filter(|s| !s.is_empty());
            let scope_version = scope
                .map(|s| s.version.clone())
                .filter(|s| !s.is_empty());
            let scope_attributes = scope
                .map(|s| key_values_to_json(&s.attributes))
                .unwrap_or_else(|| json!([]));
            let scope_dropped = scope.map(|s| s.dropped_attributes_count).unwrap_or(0);
            let scope_schema_url = None;

            for span in scope_spans.spans {
                out.push(map_span(
                    span,
                    service_name.clone(),
                    resource_attributes.clone(),
                    resource_dropped,
                    resource_schema_url.clone(),
                    scope_name.clone(),
                    scope_version.clone(),
                    scope_attributes.clone(),
                    scope_dropped,
                    scope_schema_url.clone(),
                )?);
            }
        }
    }

    Ok(out)
}

fn map_span(
    span: Span,
    service_name: String,
    resource_attributes: serde_json::Value,
    resource_dropped_attributes_count: u32,
    resource_schema_url: Option<String>,
    scope_name: Option<String>,
    scope_version: Option<String>,
    scope_attributes: serde_json::Value,
    scope_dropped_attributes_count: u32,
    scope_schema_url: Option<String>,
) -> Result<SpanRecord, OtlpError> {
    let trace_id = normalize_trace_id_bytes(&span.trace_id)
        .map_err(|e| OtlpError::Validation(e.to_string()))?;
    let span_id = normalize_span_id_bytes(&span.span_id)
        .map_err(|e| OtlpError::Validation(e.to_string()))?;

    let parent_span_id = if span.parent_span_id.is_empty() {
        None
    } else {
        Some(
            normalize_span_id_bytes(&span.parent_span_id)
                .map_err(|e| OtlpError::Validation(e.to_string()))?,
        )
    };

    let end_time_ns = if span.end_time_unix_nano == 0 {
        span.start_time_unix_nano
    } else {
        span.end_time_unix_nano
    };
    let duration_ns = end_time_ns.saturating_sub(span.start_time_unix_nano);

    let status = span.status.unwrap_or(Status {
        message: String::new(),
        code: 0,
    });

    let events = json!(span.events.iter().map(|event| {
        json!({
            "timeUnixNano": event.time_unix_nano.to_string(),
            "name": event.name,
            "attributes": key_values_to_json(&event.attributes),
            "droppedAttributesCount": event.dropped_attributes_count,
        })
    }).collect::<Vec<_>>());

    let links = json!(span.links.iter().map(|link| {
        let trace_id = normalize_trace_id_bytes(&link.trace_id).ok();
        let span_id = normalize_span_id_bytes(&link.span_id).ok();
        json!({
            "traceId": trace_id,
            "spanId": span_id,
            "traceState": link.trace_state,
            "attributes": key_values_to_json(&link.attributes),
            "droppedAttributesCount": link.dropped_attributes_count,
            "flags": link.flags,
        })
    }).collect::<Vec<_>>());

    Ok(SpanRecord {
        trace_id,
        span_id,
        parent_span_id,
        name: span.name,
        kind: span.kind,
        start_time_ns: span.start_time_unix_nano,
        end_time_ns,
        duration_ns,
        status_code: status.code,
        status_message: if status.message.is_empty() {
            None
        } else {
            Some(status.message)
        },
        trace_state: if span.trace_state.is_empty() {
            None
        } else {
            Some(span.trace_state)
        },
        flags: span.flags,
        dropped_attributes_count: span.dropped_attributes_count,
        dropped_events_count: span.dropped_events_count,
        dropped_links_count: span.dropped_links_count,
        service_name,
        resource_attributes,
        resource_dropped_attributes_count,
        resource_schema_url,
        scope_name,
        scope_version,
        scope_attributes,
        scope_dropped_attributes_count,
        scope_schema_url,
        attributes: key_values_to_json(&span.attributes),
        events,
        links,
    })
}

pub fn map_traces_json(body: &[u8]) -> Result<Vec<SpanRecord>, OtlpError> {
    let request: ExportTraceServiceRequest = serde_json::from_slice(body)
        .map_err(|e| OtlpError::InvalidPayload(format!("json decode failed: {e}")))?;
    map_traces(request)
}

pub fn map_traces_protobuf(body: &[u8]) -> Result<Vec<SpanRecord>, OtlpError> {
    let request = ExportTraceServiceRequest::decode(body)
        .map_err(|e| OtlpError::InvalidPayload(format!("protobuf decode failed: {e}")))?;
    map_traces(request)
}

// Re-export prost Message trait usage
use prost::Message;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn rejects_all_zero_trace_id_in_json() {
        let payload = br#"{"resourceSpans":[{"scopeSpans":[{"spans":[{"traceId":"00000000000000000000000000000000","spanId":"EEE19B7EC3C1B174","name":"x","startTimeUnixNano":"1","endTimeUnixNano":"2"}]}]}]}"#;
        assert!(map_traces_json(payload).is_err());
    }

    #[test]
    fn accepts_uppercase_hex_in_json() {
        let payload = br#"{"resourceSpans":[{"resource":{"attributes":[{"key":"service.name","value":{"stringValue":"demo"}}]},"scopeSpans":[{"spans":[{"traceId":"5B8EFFF798038103D269B633813FC60C","spanId":"EEE19B7EC3C1B174","name":"GET /users","kind":1,"startTimeUnixNano":"1544712660000000000","endTimeUnixNano":"1544712661000000000","status":{}}]}]}]}"#;
        let spans = map_traces_json(payload).expect("valid payload");
        assert_eq!(spans.len(), 1);
        assert_eq!(spans[0].trace_id, "5b8efff798038103d269b633813fc60c");
    }
}

File: apps/api/src/features/ingest/providers/otlp/mappers/traces.ts
import { IdError, optionalOtlpId, parseOtlpId } from "../helpers/ids"
import { toBigInt } from "../../../../../shared/helpers"
import type { Json } from "../../../../../shared/helpers"
import type { SpanRecord } from "../../../../traces/types/span"
import { IngestError } from "../../errors"
import { keyValuesToJson, serviceNameFromResource } from "../helpers/values"

function asList(value: unknown): Record<string, unknown>[] {
  return Array.isArray(value) ? (value as Record<string, unknown>[]) : []
}

function unixNano(value: unknown): bigint {
  return toBigInt(value, 0n)
}

export function mapTraceRequest(request: Record<string, unknown>): SpanRecord[] {
  const out: SpanRecord[] = []
  const resourceSpans = asList(request.resourceSpans ?? request.resource_spans)

  for (const rs of resourceSpans) {
    const resource = (rs.resource ?? {}) as Record<string, unknown>
    const resourceAttributes = keyValuesToJson(resource.attributes)
    const resourceDropped = Number(resource.droppedAttributesCount ?? resource.dropped_attributes_count ?? 0)
    const serviceName = serviceNameFromResource(resource.attributes)

    for (const ss of asList(rs.scopeSpans ?? rs.scope_spans)) {
      const scope = (ss.scope ?? {}) as Record<string, unknown>
      const scopeName = String(scope.name ?? "") || undefined
      const scopeVersion = String(scope.version ?? "") || undefined
      const scopeAttributes = keyValuesToJson(scope.attributes)
      const scopeDropped = Number(scope.droppedAttributesCount ?? scope.dropped_attributes_count ?? 0)

      for (const span of asList(ss.spans)) {
        try {
          out.push(
            mapSpan(
              span,
              serviceName,
              resourceAttributes,
              resourceDropped,
              scopeName,
              scopeVersion,
              scopeAttributes,
              scopeDropped,
            ),
          )
        } catch (err) {
          if (err instanceof IdError) {
            throw new IngestError("validation", err.message)
          }
          throw err
        }
      }
    }
  }

  return out
}

function mapSpan(
  span: Record<string, unknown>,
  serviceName: string,
  resourceAttributes: Json,
  resourceDroppedAttributesCount: number,
  scopeName: string | undefined,
  scopeVersion: string | undefined,
  scopeAttributes: Json,
  scopeDroppedAttributesCount: number,
): SpanRecord {
  const start = unixNano(span.startTimeUnixNano ?? span.start_time_unix_nano)
  const rawEnd = unixNano(span.endTimeUnixNano ?? span.end_time_unix_nano)
  const end = rawEnd === 0n ? start : rawEnd
  const status = (span.status ?? {}) as Record<string, unknown>
  const statusCode = Number(status.code ?? 0)
  const statusMessage = String(status.message ?? "")

  const events = asList(span.events).map((event) => ({
    timeUnixNano: String(event.timeUnixNano ?? event.time_unix_nano ?? "0"),
    name: String(event.name ?? ""),
    attributes: keyValuesToJson(event.attributes),
    droppedAttributesCount: Number(
      event.droppedAttributesCount ?? event.dropped_attributes_count ?? 0,
    ),
  }))

  const links = asList(span.links).map((link) => ({
    traceId: optionalOtlpId(link.traceId ?? link.trace_id, "trace") ?? null,
    spanId: optionalOtlpId(link.spanId ?? link.span_id, "span") ?? null,
    traceState: String(link.traceState ?? link.trace_state ?? ""),
    attributes: keyValuesToJson(link.attributes),
    droppedAttributesCount: Number(
      link.droppedAttributesCount ?? link.dropped_attributes_count ?? 0,
    ),
    flags: Number(link.flags ?? 0),
  }))

  const parentRaw = span.parentSpanId ?? span.parent_span_id
  const parentEmpty =
    parentRaw == null ||
    parentRaw === "" ||
    (parentRaw instanceof Uint8Array && parentRaw.length === 0)

  return {
    traceId: parseOtlpId(span.traceId ?? span.trace_id, "trace"),
    spanId: parseOtlpId(span.spanId ?? span.span_id, "span"),
    parentSpanId: parentEmpty ? undefined : parseOtlpId(parentRaw, "span"),
    name: String(span.name ?? ""),
    kind: Number(span.kind ?? 0),
    startTimeNs: start,
    endTimeNs: end,
    durationNs: end > start ? end - start : 0n,
    statusCode,
    statusMessage: statusMessage || undefined,
    traceState: String(span.traceState ?? span.trace_state ?? "") || undefined,
    flags: Number(span.flags ?? 0),
    droppedAttributesCount: Number(
      span.droppedAttributesCount ?? span.dropped_attributes_count ?? 0,
    ),
    droppedEventsCount: Number(
      span.droppedEventsCount ?? span.dropped_events_count ?? 0,
    ),
    droppedLinksCount: Number(
      span.droppedLinksCount ?? span.dropped_links_count ?? 0,
    ),
    serviceName,
    resourceAttributes,
    resourceDroppedAttributesCount,
    scopeName,
    scopeVersion,
    scopeAttributes,
    scopeDroppedAttributesCount,
    attributes: keyValuesToJson(span.attributes),
    events,
    links,
  }
}
